Scohen/encapsulate source file

scohen opened the pull request:
@lukaszsamson this is my proposal for the start of a new source file replacement. The relevant commit is 
https://github.com/elixir-lsp/elixir-ls/commit/bb866da9685c534400487a0bb810b0f719351dbe

The eventual goal would be to start passing around the struct defined by this module rather than a string to represent a source file. Users of that data structure would then operate on the API provided by `SourceFile.Experimental`. 

I see the advantages of doing so as being
  * Having a well-defined public interface
  * Centralize parsing and fetching of lines
  * Prevent O(n) queries to get context inside of a source file
  * Operate on an actual data structure whose internals can be improved to suit the needs of the project, rather than a string
  * Prevent re-parsing the same text multiple times.
  * Completely remove the need to convert between paths and URIs. The structure knows both. 

Source files would eventually "live" inside the language server genserver inside of a cache as they do now, but with a more defined interface. They would be created, modified, found and destroyed in response to messages coming in from the client, and then passed around to satisfy the providers' needs. 

The next step would be to go around and change the signatures of the providers to accept a source_file struct and operate on it. I think this would also mean that URIs don't need to be passed around any more. 

I've ported over all existing tests that dealt with content changes, and all of them pass, with the exception of those tests that were explicitly dealing with behavior that the language server protocol expressly forbids, like negative positions in documents, and I think that was one or two tests, max. 

Thoughts welcome and appreciated, and I'm creating this as a draft to encourage discussion, however, merging this as-is won't affect any existing functionality.

lukaszsamson commented on apps/language_server/lib/language_server/experimental/source_file/document.ex:
```diff
@@ -0,0 +1,73 @@
+    lines =
+      text
+      |> LineParser.parse(starting_index)
+      |> :array.from_list()
```
do `:array` make any difference here?

lukaszsamson commented:
@scohen I finally got some time to look into this PR and I have to say it's impressive. I like the direction you took. I'm going to play around with it.

> Prevent O(n) queries to get context inside of a source file

At what number of lines `:array` outperforms `List`? Last time I tried it turned out slower.

> I've ported over all existing tests that dealt with content changes, and all of them pass, with the exception of those tests that were explicitly dealing with behavior that the language server protocol expressly forbids, like negative positions in documents, and I think that was one or two tests, max.

I think we should keep those test and leave them commented out with some info

> I think this would also mean that URIs don't need to be passed around any more

That might not be easy. We only keep open files as `SourceFile` structs. There are many cases when we have to deal with not open files (e.g. LSP file watcher, mix compiler diagnostics, symbol references)

> and might need to actually go inside elixir_sense to reap the maximum benefit

I'm going to move most (all?) functionality from elixir_sense to elixirLS. I'm not aware of anyone using it and keeping it as a separate lib makes development unnecessarily complicated.

scohen commented:
Here's the benchmark I did using benchfella
```
##### With input 1500 lines #####
Name                           ips        average  deviation         median         99th %
Document                  725.59 K        1.38 μs  ±2637.77%        0.97 μs        2.68 μs
String.split/Enum.at        4.52 K      221.25 μs    ±19.65%      213.96 μs      402.69 μs
Document with build         0.47 K     2147.64 μs    ±18.17%     1986.74 μs     3677.32 μs

Comparison:
Document                  725.59 K
String.split/Enum.at        4.52 K - 160.54x slower +219.88 μs
Document with build         0.47 K - 1558.30x slower +2146.26 μs

##### With input 500 lines #####
Name                           ips        average  deviation         median         99th %
Document                  829.79 K        1.21 μs  ±2288.60%        0.87 μs        1.97 μs
String.split/Enum.at       14.52 K       68.85 μs    ±18.86%       65.96 μs      129.45 μs
Document with build         1.18 K      844.42 μs    ±34.45%      748.48 μs     1907.71 μs

Comparison:
Document                  829.79 K
String.split/Enum.at       14.52 K - 57.13x slower +67.64 μs
Document with build         1.18 K - 700.69x slower +843.22 μs

##### With input 80 lines #####
Name                           ips        average  deviation         median         99th %
Document                  879.65 K        1.14 μs  ±2593.24%        0.85 μs        1.62 μs
String.split/Enum.at      142.94 K        7.00 μs   ±132.46%        6.31 μs       25.18 μs
Document with build        22.30 K       44.85 μs    ±25.54%       42.66 μs       92.52 μs

Comparison:
Document                  879.65 K
String.split/Enum.at      142.94 K - 6.15x slower +5.86 μs
Document with build        22.30 K - 39.45x slower +43.72 μs
```

scohen commented:
One last thing: 
I've also been experimenting with using a `SourceFile.Store` genserver that holds the parsed source file. Doing this means that `elixir_ls` can have a "parse on first read" behavior and also store AST code. It would also allow for storing previous versions of the source code if that's at all helpful.

scohen commented:
@lukaszsamson While playing around with the benchmark, I noticed that protocol consolidation was disabled in elixir_ls for some reason, so I enabled it for all environments except test. This had a marked impact on performance, since I was using `Enum.at` inside of the `fetch_line`function. 
```
##### With input 1500 lines #####
Name                                          ips        average  deviation         median         99th %
Document                                   1.49 M     0.00067 ms  ±5592.81%     0.00047 ms     0.00081 ms
String.split |> Enum.at                 0.00448 M        0.22 ms    ±20.94%        0.21 ms        0.41 ms
Document.new |> Document.fetch_line     0.00046 M        2.17 ms    ±14.87%        2.06 ms        3.43 ms

Comparison:
Document                                   1.49 M
String.split |> Enum.at                 0.00448 M - 331.93x slower +0.22 ms
Document.new |> Document.fetch_line     0.00046 M - 3222.10x slower +2.17 ms

##### With input 500 lines #####
Name                                          ips        average  deviation         median         99th %
Document                                1755.42 K        0.57 μs  ±5226.87%        0.42 μs        0.74 μs
String.split |> Enum.at                   14.97 K       66.80 μs    ±26.36%       61.85 μs      146.13 μs
Document.new |> Document.fetch_line        1.11 K      901.61 μs    ±26.77%      814.98 μs     1759.53 μs

Comparison:
Document                                1755.42 K
String.split |> Enum.at                   14.97 K - 117.26x slower +66.23 μs
Document.new |> Document.fetch_line        1.11 K - 1582.70x slower +901.04 μs

##### With input 80 lines #####
Name                                          ips        average  deviation         median         99th %
Document                                2118.73 K        0.47 μs  ±4071.01%        0.38 μs        0.67 μs
String.split |> Enum.at                  154.90 K        6.46 μs   ±145.31%        5.65 μs       27.86 μs
Document.new |> Document.fetch_line       27.42 K       36.46 μs    ±47.21%       32.73 μs       95.33 μs

Comparison:
Document                                2118.73 K
String.split |> Enum.at                  154.90 K - 13.68x slower +5.98 μs
Document.new |> Document.fetch_line       27.42 K - 77.26x slower +35.99 μs
```

I pushed a change that has the benchmark and protocol consolidation.

scohen commented:
...and after doing all this benchmarking, I didn't answer the question about access. When is `:array.get` faster than `Enum.at` for a list?
I ran the following benchmark:
```
##### With input 1500 lines #####
Name                                                 ips        average  deviation         median         99th %
tuple                                             7.96 M       0.126 μs ±19437.72%       0.103 μs       0.136 μs
:array.get(count - 1, array)                      5.19 M       0.193 μs ±12840.03%       0.162 μs        0.26 μs
Enum.at(lines, count - 1)                        0.158 M        6.32 μs   ±132.51%        5.93 μs       21.34 μs
list |> List.to_tuple() |> elem(count - 1)      0.0927 M       10.79 μs   ±136.40%       10.13 μs       32.54 μs

Comparison:
tuple                                             7.96 M
:array.get(count - 1, array)                      5.19 M - 1.53x slower +0.0671 μs
Enum.at(lines, count - 1)                        0.158 M - 50.36x slower +6.20 μs
list |> List.to_tuple() |> elem(count - 1)      0.0927 M - 85.90x slower +10.66 μs

##### With input 500 lines #####
Name                                                 ips        average  deviation         median         99th %
tuple                                             9.23 M       0.108 μs  ±4245.52%       0.102 μs       0.131 μs
:array.get(count - 1, array)                      6.51 M       0.154 μs  ±3475.37%       0.142 μs       0.184 μs
Enum.at(lines, count - 1)                         0.53 M        1.88 μs   ±663.97%        1.75 μs        2.20 μs
list |> List.to_tuple() |> elem(count - 1)        0.35 M        2.89 μs   ±595.18%        2.65 μs        4.30 μs

Comparison:
tuple                                             9.23 M
:array.get(count - 1, array)                      6.51 M - 1.42x slower +0.0453 μs
Enum.at(lines, count - 1)                         0.53 M - 17.37x slower +1.77 μs
list |> List.to_tuple() |> elem(count - 1)        0.35 M - 26.63x slower +2.78 μs

##### With input 80 lines #####
Name                                                 ips        average  deviation         median         99th %
tuple                                             9.24 M      108.23 ns   ±286.39%         102 ns         132 ns
:array.get(count - 1, array)                      7.54 M      132.61 ns   ±262.15%         123 ns         170 ns
Enum.at(lines, count - 1)                         3.31 M      302.09 ns  ±2209.15%         276 ns         531 ns
list |> List.to_tuple() |> elem(count - 1)        3.10 M      323.06 ns  ±7847.13%         236 ns         439 ns

Comparison:
tuple                                             9.24 M
:array.get(count - 1, array)                      7.54 M - 1.23x slower +24.37 ns
Enum.at(lines, count - 1)                         3.31 M - 2.79x slower +193.86 ns
list |> List.to_tuple() |> elem(count - 1)        3.10 M - 2.98x slower +214.83 ns
```
From this, I believe the backing store should be a tuple.

scohen commented:
Hoo boy, I just love benchmarks. 
With an enhanced `Enumerable` implementation, switching the backing store to a tuple, enabling protocol consolidation, and doubling the performance of the line parser, here's the latest performance numbers for Document. 
Parity at small sizes and blows `Enum.at` out of the water at larger sizes. 

```
##### With input 1500 lines #####
Name                                          ips        average  deviation         median         99th %
Document                                3670.53 K        0.27 μs ±11422.26%        0.21 μs        0.33 μs
Enum.at                                  163.29 K        6.12 μs   ±102.46%        5.85 μs       18.75 μs
String.split |> Enum.at                    4.68 K      213.50 μs    ±14.68%      207.46 μs      311.07 μs
Document.new |> Document.fetch_line        0.45 K     2201.33 μs    ±23.52%     2002.04 μs     4074.42 μs

Comparison:
Document                                3670.53 K
Enum.at                                  163.29 K - 22.48x slower +5.85 μs
String.split |> Enum.at                    4.68 K - 783.66x slower +213.23 μs
Document.new |> Document.fetch_line        0.45 K - 8080.04x slower +2201.06 μs

##### With input 500 lines #####
Name                                          ips        average  deviation         median         99th %
Document                                3562.06 K        0.28 μs ±11777.60%       0.188 μs        0.31 μs
Enum.at                                  511.29 K        1.96 μs   ±308.43%        1.79 μs        2.90 μs
String.split |> Enum.at                   15.89 K       62.94 μs    ±21.88%       59.60 μs      121.90 μs
Document.new |> Document.fetch_line        1.21 K      828.30 μs    ±35.38%      726.60 μs     1822.85 μs

Comparison:
Document                                3562.06 K
Enum.at                                  511.29 K - 6.97x slower +1.68 μs
String.split |> Enum.at                   15.89 K - 224.19x slower +62.66 μs
Document.new |> Document.fetch_line        1.21 K - 2950.46x slower +828.02 μs

##### With input 80 lines #####
Name                                          ips        average  deviation         median         99th %
Document                                   3.68 M        0.27 μs ±13858.52%       0.185 μs        0.29 μs
Enum.at                                    3.09 M        0.32 μs  ±1434.33%        0.28 μs        0.59 μs
String.split |> Enum.at                   0.151 M        6.62 μs   ±211.69%        6.24 μs       19.73 μs
Document.new |> Document.fetch_line      0.0236 M       42.39 μs    ±24.98%       40.13 μs       86.71 μs

Comparison:
Document                                   3.68 M
Enum.at                                    3.09 M - 1.19x slower +0.0520 μs
String.split |> Enum.at                   0.151 M - 24.34x slower +6.35 μs
Document.new |> Document.fetch_line      0.0236 M - 155.86x slower +42.11 μs
```

lukaszsamson commented:
> Slight pushback here; Should we keep tests that test behavior that the LSP protocol expressly forbids? I haven't seen other tests like this in the source, and I think this is an instance (the client is sending malformed requests) where the server should crash on invalid input, as it indicates a grossly out of compliance client.

+1

> Then maybe we standardize by passing around URIs as opposed to paths? I'm agnostic about which is passed around, but having one thing should present a nicer interface than having two.

We need paths in calls to elixir stdlib APIs

> I found that most of the utf16 -> utf8 -> utf16 conversions can be eliminated if each line is ascii

True but can we easily check that? Anyway, I hope in the near future LSP 3.17 utf8 positions will have more client support so we can drop utf16

> I've also been experimenting with using a SourceFile.Store genserver that holds the parsed source file. Doing this means that elixir_ls can have a "parse on first read" behavior and also store AST code. It would also allow for storing previous versions of the source code if that's at all helpful.

I was also thinking of that but never got to implementing it. Elixir parser is quite fast. I wonder if we are going to run into garbage collector issues if we start passing around big trees between processes (I don't know if elixir clones binary literals or refcounts)

>  I noticed that protocol consolidation was disabled in elixir_ls for some reason

Oh, I didn't notice it. Blame tells that it's there since initial code dump. We'll need some testing to see if it not interferes with client code and release process

scohen commented:
> True but can we easily check that? 
We can! The current implementation visits every byte and checks for non-ascii characters in utf-8 streams, and I'm pretty sure we can detect if a line is utf-16 encoded by checking for lots of zero bytes (or we could just make the assumption that no utf-8 source code file will have a null character at position 0 and use that heuristic. I could also flip the current heuristic around and say that by default a line _isn't_ ascii and if every character is less than 128 and greater than 0 then it is.

> Anyway, I hope in the near future LSP 3.17 utf8 positions will have more client support so we can drop utf16
Will that be possible? I read that this is handled during negotiation, and it's possible that windows clients (it's always windows, isn't it?) will still pick utf-16. You'd know more than I though.

> I was also thinking of that but never got to implementing it. Elixir parser is quite fast. I wonder if we are going to run into garbage collector issues if we start passing around big trees between processes (I don't know if elixir clones binary literals or refcounts)

BEAM keeps a common string pool for all binaries  over 64 bytes (they're called refc binaries), so we should be good on that front. I wasn't thinking of it as a performance improvement so much as a correctness / capabilities improvement. It would allow us to do things like line-by-line parsing, realizing AST when needed and handling multiple requests with a single action. 

> Oh, I didn't notice it. Blame tells that it's there since initial code dump. We'll need some testing to see if it not interferes with client code and release process
 
You should be good here with minimal testing --enabling in test mode should do it. Consolidation just makes things faster. If this was really old, it used to take a long time during builds which might be why it was disabled.

scohen commented:
@lukaszsamson I made one more change; I implemented doing character conversion at the start of the request, which re-encodes the horribly named `character` key of the position to the byte offset of the utf-16 character... but only if the line isn't ascii; I added some tests around `LineParser` parsing utf-16 and ensuring it flags that as "not ascii", and finally, fixed a bug in conversions that was looking at the wrong line when deciding if we should renumber a change or not. 

The result is that we now don't do any utf8 -> utf16 -> utf8 conversions unless we need to, and the decision is line-by-line. I really like the simplified code in `source_file.ex` now.

lukaszsamson commented on apps/language_server/test/experimental/source_file_test.exs:
```diff
@@ -0,0 +1,771 @@
+      assert "foooo\nbaz🏳️‍🌈z\nbaz" == text(source)
+    end
+
+    test "invalid update range - before the document starts -> before the document starts" do
```
I think we should keep the behavior aligned with https://github.com/microsoft/vscode-languageserver-node/blob/df9ea01af578e02d910bf0fb4c7d4edcd94d4e55/textDocument/src/test/textdocument.test.ts#L379

scohen replied:
I'll do it, but those seem like extremely poor tests to me, they expressly violate the spec for what the position can contain
(https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#position says both line and character are uinteger which is defined here https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#uinteger as having a range from 0 to 2^31 - 1)
and they introduce undefined behavior. Nowhere in the spec does it say that positions can be negative (in fact, it says the opposite) and it also doesn't say what to do with them if they are. Seeing negative positions indicates a wildly out of spec client.

I'll push a commit, but I beg you to reconsider. Following a spec is better than following microsoft's slapdash tests.

lukaszsamson replied:
Oh, nice catch @scohen , I missed that the type is `uinteger`. Then let's drop that tests. The only specified behavior is for `character` greater than line length.

lukaszsamson commented on apps/language_server/test/experimental/source_file_test.exs:
```diff
@@ -0,0 +1,775 @@
+    test "elixir_position_to_lsp multi line" do
+      assert {1, 1} == SourceFile.elixir_position_to_lsp("abcde\n1234", {2, 2})
+    end
+
```
Please add this test https://github.com/elixir-lsp/elixir-ls/blob/f02c97a7ba4631d91b0c18f0fd4aac1f8861427f/apps/language_server/test/source_file_test.exs#L665

lukaszsamson replied:
And consider also this https://github.com/elixir-lsp/elixir-ls/blob/f02c97a7ba4631d91b0c18f0fd4aac1f8861427f/apps/language_server/test/source_file_test.exs#L684

lukaszsamson commented:
@scohen I think it's almost ready to be merged. Please revert the changes with negative positions handling and add a test for `character` greater than line length (and possibly `line` greater than number of lines - 1). BTW CI on master should be more stable now.

scohen commented:
@lukaszsamson I pushed all the changes you asked for.
